dll/win32/shlwapi/propbag.cpp

```diff
@@ -251,3 +251,323 @@ SHCreatePropertyBagOnMemory(_In_ DWORD dwMode, _In_ REFIID riid, _Out_ void **pp
 
     return hr;
 }
+
+class CRegPropertyBag : public CBasePropertyBag
+{
+protected:
+    HKEY m_hKey;
+
+    HRESULT _ReadDword(LPCWSTR pszPropName, VARIANT *pvari);
+    HRESULT _ReadString(LPCWSTR pszPropName, VARIANTARG *pvarg, UINT len);
+    HRESULT _ReadBinary(LPCWSTR pszPropName, VARIANT *pvari, VARTYPE vt, DWORD uBytes);
+    HRESULT _ReadStream(VARIANT *pvari, BYTE *pInit, UINT cbInit);
+    HRESULT _CopyStreamIntoBuff(IStream *pStream, void *pv, ULONG cb);
+    HRESULT _GetStreamSize(IStream *pStream, LPDWORD pcbSize);
+    HRESULT _WriteStream(LPCWSTR pszPropName, IStream *pStream);
+
+public:
+    CRegPropertyBag(DWORD dwMode)
+        : CBasePropertyBag(dwMode)
+        , m_hKey(NULL)
+    {
+    }
+
+    ~CRegPropertyBag() override
+    {
+        if (m_hKey)
+            ::RegCloseKey(m_hKey);
+    }
+
+    HRESULT Init(HKEY hKey, LPCWSTR lpSubKey);
+
+    STDMETHODIMP Read(_In_z_ LPCWSTR pszPropName, _Inout_ VARIANT *pvari,
+                      _Inout_opt_ IErrorLog *pErrorLog) override;
+    STDMETHODIMP Write(_In_z_ LPCWSTR pszPropName, _In_ VARIANT *pvari) override;
+};
+
+HRESULT CRegPropertyBag::Init(HKEY hKey, LPCWSTR lpSubKey)
+{
+    REGSAM nAccess = 0;
+    if ((m_dwMode & (STGM_READ | STGM_WRITE | STGM_READWRITE)) != STGM_WRITE)
+        nAccess |= KEY_READ;
+    if ((m_dwMode & (STGM_READ | STGM_WRITE | STGM_READWRITE)) != STGM_READ)
+        nAccess |= KEY_WRITE;
+
+    LONG error;
+    if (m_dwMode & STGM_CREATE)
+        error = ::RegCreateKeyExW(hKey, lpSubKey, 0, NULL, 0, nAccess, NULL, &m_hKey, NULL);
+    else
+        error = ::RegOpenKeyExW(hKey, lpSubKey, 0, nAccess, &m_hKey);
+
+    if (error != ERROR_SUCCESS)
+        return (LOWORD(error) | 0x80070000); // FIXME: What is 0x80070000?
```

Comment:
Do you get the 8070000 in WinXP or 2k3 ? Or in Win10 ?

Reply:
Got in WinXP, Win2k3, and Win10.